explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,153 @@
+
+```bash
+$ sudo perf record \
+	-e br_inst_retired.near_taken \
+	--branch-any \
+	--all-cpus \
+	--output openssl-nonpgo.perfdata \
+	-- openssl speed -mr -seconds 60 -evp md5 sha512 rsa2048 aes-256-cbc
+```
+
+On an AMD system, use the following invocation:
+
+```bash
+$ sudo perf record \
+	-e 'cpu/event=0xc4,umask=0x0,name=ex_ret_brn_tkn/' \
+	--branch-any \
+	--all-cpus \
+	--output openssl-nonpgo.perfdata \
+	-- openssl speed -mr -seconds 60 -evp md5 sha512 rsa2048 aes-256-cbc
+```
+
+Note that the only thing that differs is the PMU event to be monitored.
+
+After the command has finished running, you should see a file named `openssl-nonpgo.perfdata` in your current directory.  This is the file we will now need to convert to `gcov`.
+
+### Converting `perfdata` to `gcov` with `autofdo`
+
+With `autofdo` installed, you can convert the `perfdata` generated in the last step to `gcov` by doing:
+
+```bash
+$ sudo chown ubuntu:ubuntu openssl-nonpgo.perfdata
+$ create_gcov \
+	--binary /usr/bin/openssl \
+	--gcov openssl-nonpgo.gcov \
+	--profile openssl-nonpgo.perfdata \
+	--gcov_version 2 \
+	-use_lbr false
+```
+
+`create_gcov` is verbose and will display several messages that may look like something is wrong; you should inspect its output and also make sure to check the `$?` shell variable to make sure that it has finished successfully.  You will then have a file called `openssl-nonpgo.gcov` in your current directory, which is the `gcov` file we will use to feed back to GCC when recompiling OpenSSL.  You can inspect the file by using the `dump_gcov` tool.
+
+### Rebuilding OpenSSL with PGO
+
+We have everything we need to rebuild OpenSSL and make use of our profile data.  The most important thing now is to set the `CFLAGS` environment variable properly so that GCC can find and use the profile we generated in the previous step.
+
+How you perform this `CFLAGS` adjustment depends on how you built the software in the first place, so we won't cover this part here.  The resulting `CFLAGS` variable should have the following GCC option, though:
+
+```
+-fauto-profile=/path/to/openssl-nonpgo.gcov
+```
+
+Make sure to adjust the path to the `openssl-nonpgo.gcov` file.
+
+Also note that enabling PGO will make GCC automatically enable several optimization passes that are usually disable, even when using `-O2`.  This can lead to new warnings/errors in the code, which might require either the code to be adjusted or the use of extra compilation flags to suppress these warnings.
+
+After the software is built, you can install this new version and run the benchmark tests again to compare results.
+
+### Our results
+
+We obtained some interesting results while performing the tests.  Here are the highlights:
+
+* Overall, the performance gain/loss was minimal.  Most of the time it was less than 1%.
```

Comment:
That as the first message is a bit non-inspiring.

We might want to mention that this is due to these hot code paths already being some that are highly optimized (they are) and that you've found with more complex software (in your other tests much more than a hand optimized hot loop is active) you've seen usually 5%-15%

Without framing it this way of "we chosen this for simplicity of the test, but thereby accepting we won't see much - yet you can have much we are shooting the messaging and helpfulness in the foot".

Once the blog exists which is more reporting what we did vs showing how one could recreate the same you can even point to it from here.

Reply:
If someone is following along, won't their results depend on their hardware? If you're presenting a blog with the results of your testing, I agree that those should go here as an example of what could be seen. Since during the article you're using we/our throughout, it's not necessarily clear that "our results" here actually refers to the results of your testing

Reply:
@cpaelzer Good suggestion.  I'll add some text expanding on the reason why we've seen such a low performance gain.

@s-makin The results will depend on a bunch of things, including their hardware, yes.  I still have to prepare the blog post, but I'll certainly link it here once it's done.